crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1107,6 +1123,104 @@ impl<'db> TypeInferenceBuilder<'db> {
+
+        if let Some(declared_ty) = function
+            .returns
+            .as_deref()
+            .map(|ret| self.file_expression_type(ret))
+        {
+            fn is_stub_suite(suite: &[ast::Stmt]) -> bool {
+                match suite {
+                    [ast::Stmt::Expr(ast::StmtExpr { value: first, .. }), ast::Stmt::Expr(ast::StmtExpr { value: second, .. }), ..] => {
+                        first.is_string_literal_expr() && second.is_ellipsis_literal_expr()
+                    }
+                    [ast::Stmt::Expr(ast::StmtExpr { value, .. }), ast::Stmt::Pass(_), ..] => {
+                        value.is_string_literal_expr()
+                    }
+                    [ast::Stmt::Expr(ast::StmtExpr { value, .. }), ..] => {
+                        value.is_ellipsis_literal_expr()
+                    }
+                    [ast::Stmt::Pass(_)] => true,
+                    _ => false,
+                }
+            }
+            fn does_suite_always_return(suite: &[ast::Stmt]) -> bool {
+                match suite.last() {
+                    Some(ast::Stmt::Return(_) | ast::Stmt::Raise(_)) => true,
+                    Some(ast::Stmt::If(if_stmt)) => {
+                        !if_stmt.elif_else_clauses.is_empty()
+                            && does_suite_always_return(&if_stmt.body)
+                            && if_stmt
+                                .elif_else_clauses
+                                .iter()
+                                .all(|clause| does_suite_always_return(&clause.body))
+                    }
+                    Some(ast::Stmt::While(while_stmt)) => {
+                        does_suite_always_return(&while_stmt.body)
+                            && does_suite_always_return(&while_stmt.orelse)
+                    }
+                    Some(ast::Stmt::For(for_stmt)) => {
+                        does_suite_always_return(&for_stmt.body)
+                            && does_suite_always_return(&for_stmt.orelse)
+                    }
+                    Some(ast::Stmt::Try(try_stmt)) => {
+                        does_suite_always_return(&try_stmt.body)
+                            && does_suite_always_return(&try_stmt.orelse)
+                            && does_suite_always_return(&try_stmt.finalbody)
+                            && try_stmt.handlers.iter().all(|handler| {
+                                does_suite_always_return(&handler.as_except_handler().unwrap().body)
+                            })
+                    }
+                    Some(ast::Stmt::With(with_stmt)) => does_suite_always_return(&with_stmt.body),
+                    Some(ast::Stmt::Match(match_stmt)) => match_stmt
+                        .cases
+                        .iter()
+                        .all(|case| does_suite_always_return(&case.body)),
+                    _ => false,
+                }
+            }
+            let is_overloaded = function.decorator_list.iter().any(|decorator| {
```

Comment:
nit: what this is really checking is whether this function is an overload, not whether it is overloaded
```suggestion
            let is_overload = function.decorator_list.iter().any(|decorator| {
```